### DISCUSSION TITLE
How to get sighash_type from a transaction?

### DISCUSSION THREAD

**Discussion by iota9star** [rust-bitcoin/rust-bitcoin#2842] [Q&A]
Here is my code, but I'm not sure if it's correct.
```rust
use bitcoin::sighash::EcdsaSighashType;
use bitcoin::TxIn;

pub fn extract_sighash_type(vin: &TxIn) -> Option<u8> {
  let try_sighash_value = |bytes: &[u8]| {
    bytes
      .last()
      .copied()
      .map(|x| match EcdsaSighashType::from_standard(x as u32) {
        Ok(v) => Some(v.to_u32() as u8),
        Err(_) => None,
      })
      .flatten()
  };
  if !vin.witness.is_empty() {
    for x in vin.witness.iter() {
      if x.is_empty() {
        continue;
      }
      return try_sighash_value(x);
    }
  }
  if !vin.script_sig.is_empty() {
    let script_sig = &vin.script_sig;
    return try_sighash_value(script_sig.as_bytes());
  }
  None
}
```

**Comment by storopoli** [comment]
What if you directly converted what you have into a `Psbt` and use [`Psbt.sighash_ecdsa`](https://docs.rs/bitcoin/latest/bitcoin/psbt/struct.Psbt.html#method.sighash_ecdsa)?

There's also the [`Psbt.sighash_taproot`](https://docs.rs/bitcoin/latest/src/bitcoin/psbt/mod.rs.html#559) but somehow that's private and not exported.

PS: We should move this into a [Discussion](https://github.com/rust-bitcoin/rust-bitcoin/discussions)

**Comment by junderw** [comment]
Please move this to a discussion.

You would need to search for signatures, this will create tons of false positives.

**Comment by apoelstra** [comment]
You may want to use the [Miniscript interpreter API](https://docs.rs/miniscript/latest/miniscript/interpreter/index.html).

You create an `Interpreter` object then call `iter` on it. This API is not super well documented because AFAIK it is only used for debugging purposes and for weird one-off blockchain forensics. If you have a real use case, we would be quite interested in that. But could you open an issue in rust-miniscript about it?

**Comment by iota9star** [comment]
> What if you directly converted what you have into a `Psbt` and use [`Psbt.sighash_ecdsa`](https://docs.rs/bitcoin/latest/bitcoin/psbt/struct.Psbt.html#method.sighash_ecdsa)?
> 
> There's also the [`Psbt.sighash_taproot`](https://docs.rs/bitcoin/latest/src/bitcoin/psbt/mod.rs.html#559) but somehow that's private and not exported.
> 
> PS: We should move this into a [Discussion](https://github.com/rust-bitcoin/rust-bitcoin/discussions)

I don't know how to use this API.

**Comment by iota9star** [comment]
@apoelstra I have already created the corresponding issue. https://github.com/rust-bitcoin/rust-miniscript/issues/690

**Comment by sanket1729** [comment]
@iota9star what you have is correct if all inputs have simple checksig script. There is no easy way to get this directly with a simple API.

**Comment by iota9star** [comment]
@sanket1729 I don't know much about this, and the code above only gives me rough results.

**Comment by sanket1729** [comment]
@iota9star, the sighash_type is not per transaction, it is per signature. What you need is 
- Extract all signatures from a transaction. This really depends on what types of transaction redeem/witness script is. If it is a miniscript, then rust-miniscript can help here. If it is a custom exotic script, you would need a full interpreter to parse them. 
What are the witness/redeem scripts for transactions that you are using? If they are not custom/exotic you can use miniscript  API. 
- Get the sighash_type from each signature. It is possible that in the same transaction, different signatures have different sighash_types.
